### DISCUSSION TITLE
HSL Brightness/Saturation/Hue manipulation

### DISCUSSION THREAD

**Discussion by idanya** [libvips/libvips#2486] [Q&A] [answered]
I understand that libvips doesn't support HSL colorspace and the closest colorspace available is LCH. 

I'm trying to migrate code that does brightness, saturation, and Hue manipulation to use libvips with no success. 
Seems like multiplying LCH with the same values as I did with HSL doesn't get to the same result. 

I'm using CLI for the POC and this is what I'm trying: 

```
vips colourspace input.jpg output.v lch
vips linear output.v output.jpg "<brightness> <saturation> 1" "0 0 <hue>"
```

I've also compared values with GIMP output and it doesn't match. In GIMP using the menu option `Colors -> Hue-Saturation...`. I can only assume this option in GIMP is using the HSL way because I get the same result with my old software.

Could be that I'm totally not on the right path, but, Is there a way to somehow "convert" values that were intended to use as multipliers/additions on HSL to LCH? Maybe some way to use HSL with libvips?

**Comment by jcupitt** [comment, marked as answer]
Hi @idanya,

libvips has HSV, which is pretty close to HSL. Would that work?

For LCh, try for example:

```
vips colourspace test-image.jpg x.v
vips linear x.v x.jpg "1 1 1" "0 0 45"
```

To add 45 degrees to the hue angle (1/8th of a full colour wheel). It won't match HSL very closely, since LCh is much more accurate.

**Reply by idanya** [reply]
You are right, it's not close.

With HSV it's the same? add the angle to additions array at the 3rd place? because I've tried it and still getting a far result from the original. 
I've tried to test it with an [external tool](https://pinetools.com/adjust-hue-image) which gives the result I'm looking for. Setting the HUE to X  in that tool doesn't give the same result as I change to HSV and add X to the 3rd element in the additions array.

This is the command I'm using:
```
vips colourspace INPUT OUTPUT_HSV hsv
vips linear OUTPUT_HSV OUTPUT "1 1 1" "${HUE} 0 0"
```

* Don't know what this external tool does inside, it's only an example that I've found. GIMP gives the same result.

**Reply by jcupitt** [reply]
HSV is an 8-bit format, so the three channels are 0 - 255. For hue, you'll need to add (45 * 360 / 256).

I tried in Gimp 2.10 and libvips LCh adjust seems to be the same as Gimp's Colours / Hue-Chroma tool.

Gimp has another menu item called (confusingly) Colours / Hue-Saturation, which looks perhaps more HSV-ish.

**Reply by jcupitt** [reply]
Oh, and you'll need to do "clock arithmetic" on hue. In something like Python:

```python
x = ... input image
hue = x[0]
hue += 45 * 360 / 256
# hue will now probably have values over 255 .. these need to be wrapped around
hue = (hue > 255).ifthenelse(hue - 256, hue)
x = hue.bandjoin[x:1:])
```

**Reply by jcupitt** [reply]
Thinking about it, of course you can just AND with 0xff after the add. 

```python
x = ... image
x += [45 * 360 / 256, 0, 0]
x &= 0xff
```

**Reply by idanya** [reply]
Thank you for taking the time to explain in detail.
Tried your params exactly and I still cannot get to the same result.

This is the original:

![orange](https://user-images.githubusercontent.com/10362893/138606667-3a966d2c-d1ed-4d2b-915c-991dc8cfa8cd.jpg)


Using GIMP Hue-Saturation tool, I've set Hue to 45 degrees. Output:

![Screen Shot 2021-10-24 at 20 51 57](https://user-images.githubusercontent.com/10362893/138606520-03c4d235-806a-4175-b914-be7af541cbac.png)

45 * 360 / 256 = 63.28125

```
vips colourspace orange.jpg orange.jpg.temp.v hsv
vips linear orange.jpg.temp.v orange.jpg.out.jpg '1 1.1 1.1' '63 0 0'
```
output:
![Screen Shot 2021-10-24 at 20 52 27](https://user-images.githubusercontent.com/10362893/138606485-cb328f9e-fee6-45b4-a61f-81d211b2cafa.png)

**Reply by jcupitt** [reply]
Sorry, my fault (duh), try "45 * 256 / 360". I always get that reversed.

**Reply by jcupitt** [reply]
Also, it's worth remembering that, unlike LCh, HSV is not a formally defined colourspace with a standard. Every program that implements HSV is likely to do it slightly differently, and you can't really expect programs to match output.

Having said that, I see:

```
$ vips colourspace blue-orange.jpg x.v hsv
$ vips linear x.v x2.v "1 1 1" "$(bc -l <<< '45 * 256 / 360') 0 0"
$ vips boolean_const x2.v x3.v and 255
$ vips colourspace x3.v x.jpg srgb
```

To make:

![x](https://user-images.githubusercontent.com/580843/138679841-5bcc8ef6-46d9-414c-adbf-4122312e75eb.jpg)

Which looks close to GIMP.

**Reply by jcupitt** [reply]
Final post ... the libvips CLI is slow and ugly for this kind of thing. In Python it would be:

```python
#!/usr/bin/python3

import sys
import pyvips

angle = float(sys.argv[3])

x = pyvips.Image.new_from_file(sys.argv[1], access="sequential")
old_colourspace = x.interpretation
x = x.colourspace("hsv")
x = (x + [angle * 256 / 360, 0, 0]) & 0xff
x = x.colourspace(old_colourspace)
x.write_to_file(sys.argv[2])
```

Ruby / php / C++ etc. are very similar. I see:

```
$ time ./hue.py blue-orange.jpg x.jpg 45
real	0m0.224s
user	0m0.927s
sys	0m0.129s
```

So, it's easy to read (imo), and performs reasonably well.

**Reply by idanya** [reply]
That is much better and closer to the desired result. 
still not identical of course but I guess it's close enough considering the difference between colorspaces. 

Thanks a lot!

**Reply by idanya** [reply]
Last question ... 

I've managed to get a good result but I still have an issue with normalizing negative numbers. 
My input is `-180 < x < 180`, so my calculation is

1. x = (x + 360) * 256 / 360
2. if x > 255 -> x-= 256

All positive values look great, negative ones look all the same. it's like any rotation behind 180 is meaningless.

**Reply by jcupitt** [reply]
You'd need to post a test program, but I'd use `%` (remainder after division) to implement clock arithmetic.

**Reply by idanya** [reply]
Getting this part of the program out can be hard as it is a small part integrated into a much larger mechanism. 

use the remainder where?

**Reply by idanya** [reply]
`-180 < floatHue < 180`

```
floatHue = math.Mod(floatHue + 360, 360)
h = int(floatHue * 256 / 360)
if h > 255 {
	h -= 256
}
```

**Reply by jcupitt** [reply]
In Python it'd be something like:

```
# normalize to 0 - 359
hue = (hue + 360) % 360

# rescale to 0 - 255
hue *= 256.0 / 360.0
```

ie. get rid of any negatives (you could even add 720 to be really safe), then take the remainder after division.

**Reply by idanya** [reply]
*updated with +720*

Wrote a small python example 

command example:

```
script.py input.png output.png 160 55 44
```

160 = hue = values between -180 to 180
55 = saturation = values between -100 to 100 
44 brightness = values between -100 to 100

```python
#!/usr/bin/python3

import sys
import pyvips

in_file = sys.argv[1]
out_file = sys.argv[2]
hue_angle = float(sys.argv[3])
saturation = (float(sys.argv[4]) + 100) * 0.01
brightness = (float(sys.argv[5]) + 100) * 0.01

hue_angle = (hue_angle + 720) % 360
hue_angle *= 256.0 / 360.0

if hue_angle > 255:
    hue_angle -= 256

print(f"in_file: {in_file}, out_file: {out_file}, hue: {hue_angle}, sat: {saturation} br: {brightness}")

x = pyvips.Image.new_from_file(in_file, access="sequential")
old_colourspace = x.interpretation
x = x.colourspace("hsv")
x = x * [1, saturation, brightness, 1]
x = x + [hue_angle, 0, 0, 0]
x = x.colourspace(old_colourspace)
x.write_to_file(out_file)
```

**Reply by idanya** [reply]
Ok, after playing with it some more I think I've managed to find the solution. 
Looks like HSV is 256 that ranging between -128 to +128.
rescaling to 256 and leaving negative numbers be, seems to work:

Just this:
`hue_angle *= 256.0 / 360.0`

inputs: -180 < x < +180
outputs:  -128 < x < +128

Thanks a lot!

**Comment by jcupitt** [comment]
nip2 is handy for developing things like this. 

https://github.com/libvips/nip2

You can connect up all the libvips operations, change values and watch images go crazy. For example:

![x](https://user-images.githubusercontent.com/580843/137812659-67461027-0407-475a-af4e-7a7d949f8c04.png)

**Comment by alexandrearc** [comment]
Adding a comment here because I fill it is very similar, since modulate is also to change Brightness, Saturation and Hue.

How can i convert this command from ImageMagick?

`magick source.png -colorspace srgb -modulate '120,10,10' -fill '#222b6d' destination.jpeg`

**Reply by jcupitt** [reply]
Reading the modulate docs, the first two numbers are percent multiplications, and the last is clock arithmetic on hue. So your command is doing something like:

```python
brightness *= 1.2
saturation *= 0.1
hue -= 162 # degrees
```

libvips HSV is hue, saturation, value, with values in 0-255. You could write:

```
vips colourspace k2.jpg t1.v hsv
vips linear t1.v result.jpg "1 0.1 1.2" "141 0 0"
```

If you use eg. Python, you don't need the temporary file.

```python
image = image.colourspace("hsv") * [1, 0.1, 1.2] + [141, 0, 0]
```

**Reply by alexandrearc** [reply]
Thanks it worked, as far I can see.
